cachi2/core/package_managers/yarn/locators.py

```diff
@@ -1,115 +1,273 @@
+            raise UnsupportedFeature(
+                f"Cachi2 does not support Git or Exec dependencies for Yarn Berry: {locator_str}",
+                docs=None,  # TODO: docs needed
+            )
+        elif protocol == "npm":
+            return NpmLocator(locator.scope, locator.name, version=parsed_reference.selector)
+        elif protocol == "workspace":
+            relpath = Path(parsed_reference.selector)
+            return WorkspaceLocator(locator.scope, locator.name, relpath)
+        elif protocol == "patch":
+            return _parse_patch_locator(locator)
+        elif protocol in ("file", "link", "portal"):
+            return _parse_file_locator(locator)
+        elif protocol in ("http", "https") and TARBALL_RE.match(locator.raw_reference):
+            return HttpsLocator(url=locator.raw_reference)
+    except UnexpectedFormat as e:
+        raise UnexpectedFormat(f"parsing {locator_str!r}: {e}") from e
+
+    raise UnexpectedFormat(f"parsing {locator_str!r}: unknown protocol")
+
+
+def _parse_patch_locator(locator: "_ParsedLocator") -> PatchLocator:
+    # https://github.com/yarnpkg/berry/blob/b6026842dfec4b012571b5982bb74420c7682a73/packages/plugin-patch/sources/patchUtils.ts#L13
+    reference = locator.parsed_reference
+    if not reference.source:
+        raise UnexpectedFormat("missing source in locator")
+
+    original_package = parse_locator(reference.source)
+
+    # https://github.com/yarnpkg/berry/blob/b6026842dfec4b012571b5982bb74420c7682a73/packages/plugin-patch/sources/patchUtils.ts#L92
+    def process_patch_path(patch: str) -> Union[str, Path]:
+        patch = patch.removeprefix("~")  # '~' denotes an optional patch
+        if re.match(r"^builtin<([^>]+)>$", patch):
+            return patch
+        else:
+            return Path(patch)
+
+    patches = [process_patch_path(p) for p in reference.selector.split("&")]
+    if locator_param := _get_param(reference, "locator"):
+        parent_locator = parse_locator(locator_param)
+    else:
+        parent_locator = None
+
+    return PatchLocator(original_package, patches, parent_locator)
+
+
+def _parse_file_locator(locator: "_ParsedLocator") -> FileLocator:
+    reference = locator.parsed_reference
+
+    relpath = Path(reference.selector)
+    # for 'file:' directories, Yarnberry uses the path as both source and selector
+    # https://github.com/yarnpkg/berry/blob/b6026842dfec4b012571b5982bb74420c7682a73/packages/plugin-file/sources/fileUtils.ts#L16
+    if reference.source and Path(reference.source) != relpath:
+        raise UnexpectedFormat("conflicting paths in locator")
+
+    locator_param = _get_param(reference, "locator")
+    if not locator_param:
+        raise UnexpectedFormat("missing 'locator' param")
+
+    parent_locator = parse_locator(locator_param)
+    return FileLocator(relpath, parent_locator)
+
+
+def _get_param(reference: "_ParsedReference", param_name: str) -> Optional[str]:
```

Comment:
Wdyt about making this a method in _ParsedReference? It looks like a utility method to me.

Reply:
yup, that looks better :+1: